### PR TITLE
Add nvidia bert example to orttraining_rc1 //build branch

### PR DESCRIPTION
**Pull request by suffiank**
Initial add of nvidia-bert example for //build.

### CODE REVIEW

File: orttraining/pytorch_frontend_examples/nvidia-bert/Onnxrt.dockerfile [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+# below is similar to NVIDIA dockerfile
+# --- 
+
+RUN apt-get update && apt-get install -y pbzip2 pv bzip2 cabextract
+
+ENV BERT_PREP_WORKING_DIR /workspace/bert/data
+
+WORKDIR /workspace
+RUN git clone https://github.com/attardi/wikiextractor.git
+RUN git clone https://github.com/soskek/bookcorpus.git
```
Comment (SherlockNoMad) [inline orttraining/pytorch_frontend_examples/nvidia-bert/Onnxrt.dockerfile:57]: Please verify with Manash and Kaarthik. 
we might not want to use this in our dgx2

File: orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+from onnxruntime.capi.ort_trainer import LossScaler
+import torch
+
+def setup_onnxruntime_with_mpi(args):
+    from mpi4py import MPI
+    comm = MPI.COMM_WORLD
+    args.local_rank = comm.Get_rank()
+    args.world_rank = comm.Get_rank()
+    args.world_size=comm.Get_size()
+    torch.cuda.set_device(args.local_rank)
```
Comment (Tixxx) [inline orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py:12]: it's a bug here. Get_rank gives you the world rank. we need to do set_device(args.local_rank % num_gpus_locally)

File: orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            return {"alpha": 0.9, "beta": 0.999, "lambda": 0.01, "epsilon": 1e-6}
+
+    # we request ORTTrainer to create a LambOptimizer with given optimizer_attributes. 
+    # train_step does forward, backward, and optimize step.
+    model = ORTTrainer(model, None, bert_model_description(args), "LambOptimizer", 
+        map_optimizer_attributes,
+        IODescription('Learning_Rate', [1,], torch.float32),
+        device, postprocess_model=postprocess_model, 
+        gradient_accumulation_steps=args.gradient_accumulation_steps,
+        world_rank=args.local_rank, world_size=args.world_size,
```
Comment (Tixxx) [inline orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py:90]: world_rank=args.world_rank

Comment (suffiank) [reply orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py]: Updated to use args.world_rank.

File: orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py (hunk 2 above)
Comment (suffiank) [reply orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py:12]: Updated with args.local_rank = comm.Get_rank() % torch.cuda.device_count()

File: orttraining/pytorch_frontend_examples/nvidia-bert/Onnxrt.dockerfile (hunk 1 above)
Comment (suffiank) [reply orttraining/pytorch_frontend_examples/nvidia-bert/Onnxrt.dockerfile]: Send them a message asking about it. But also commented out for now.

File: orttraining/pytorch_frontend_examples/nvidia-bert/LICENSE [hunk 4]
```diff
@@ -0,0 +1,202 @@
+
+                                 Apache License
```
Comment (skaarthik) [inline orttraining/pytorch_frontend_examples/nvidia-bert/LICENSE:2]: Would it be better to replace this file with a pointer to the LICENSE file in the original repo?

File: orttraining/pytorch_frontend_examples/nvidia-bert/NOTICE [hunk 5]
```diff
@@ -0,0 +1,5 @@
+BERT PyTorch
```
Comment (skaarthik) [inline orttraining/pytorch_frontend_examples/nvidia-bert/NOTICE:1]: This file could also be replaced with a pointer or with content that is more applicable for this repo.

File: orttraining/pytorch_frontend_examples/nvidia-bert/Nvidia.dockerfile [hunk 6]
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2020 NVIDIA CORPORATION. All rights reserved.
```
Comment (skaarthik) [inline orttraining/pytorch_frontend_examples/nvidia-bert/Nvidia.dockerfile:1]: Is this file needed at all given that we have added Onnxrt.dockerfile?

File: orttraining/pytorch_frontend_examples/nvidia-bert/Onnxrt.dockerfile [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ENV PATH /usr/local/nvidia/bin:/usr/local/cuda/bin:/opt/cmake/bin:/opt/miniconda/bin:${PATH}
+ENV LD_LIBRARY_PATH /opt/miniconda/lib:$LD_LIBRARY_PATH
+
+RUN apt update -y && pip install --upgrade pip
+
+# install pytorch stable using CUDA 10.1
+RUN pip install torch==1.5.0+cu101 torchvision==0.6.0+cu101 -f https://download.pytorch.org/whl/torch_stable.html
+
+# add pytorch patch for opset 10
+COPY pyt_patch /tmp/pyt_patch
```
Comment (skaarthik) [inline orttraining/pytorch_frontend_examples/nvidia-bert/Onnxrt.dockerfile:13]: Is this a reproducible step if someone wants to build a customer Docker image derived from this?

File: orttraining/pytorch_frontend_examples/nvidia-bert/README.md [hunk 8]
```diff
@@ -0,0 +1,946 @@
+# BERT For PyTorch
```
Comment (skaarthik) [inline orttraining/pytorch_frontend_examples/nvidia-bert/README.md:1]: Let's discuss pointer vs. inclusion in repo for this file also.

**Comment by skaarthik** [comment]
If it is not too much work, let's consider removing code/contents not relevant to ORT training sample (for example, inferencing related code, empty folders with .keep files). Since we are copying the files instead of submodule or branch clone, keeping it minimal will help with code maintenance.

File: orttraining/pytorch_frontend_examples/nvidia-bert/Nvidia.dockerfile (hunk 6 above)
Comment (suffiank) [reply orttraining/pytorch_frontend_examples/nvidia-bert/Nvidia.dockerfile:1]: It was intended to run the original NVIDIA flow. I think we have since decided we only want to focus on ORT flow. So we can remove it. And possibly other files not relevant to ORT flow.

File: orttraining/pytorch_frontend_examples/nvidia-bert/LICENSE (hunk 4 above)
Comment (thiagocrepaldi) [reply orttraining/pytorch_frontend_examples/nvidia-bert/LICENSE:2]: I am not sure whether this is allowed. Apache license allows anyone performs any change to/distribution of the original software as long as the original license and copyright are "included". Maybe "pointer" is not considered "inclusion" of license.

If any change is performed to the original source code, we also have to list them

File: orttraining/pytorch_frontend_examples/nvidia-bert/NOTICE (hunk 5 above)
Comment (thiagocrepaldi) [reply orttraining/pytorch_frontend_examples/nvidia-bert/NOTICE:1]: From [section 4.d](https://www.apache.org/licenses/LICENSE-2.0): If the Work includes a "NOTICE" text file as part of its distribution, then any Derivative Works that You distribute must include a readable copy of the attribution notices contained within such NOTICE file, excluding those notices that do not pertain to any part of the Derivative Works, in at least one of the following places: within a NOTICE text file distributed as part of the Derivative Works; within the Source form or documentation, if provided along with the Derivative Works; or, within a display generated by the Derivative Works, if and wherever such third-party notices normally appear. The contents of the NOTICE file are for informational purposes only and do not modify the License. You may add Your own attribution notices within Derivative Works that You distribute, alongside or as an addendum to the NOTICE text from the Work, provided that such additional attribution notices cannot be construed as modifying the License.

**Comment by thiagocrepaldi** [comment]
> 
> 
> > Is it possible to use nvidia-bert as git submodule and prevent copying the whole source code from nvidia-bert? That could even prevent licensing issues on redistributing thrid-party source-code
> 
> Yes. This had been the original idea. But after discussion it was settled to use a folder for now. I believe the final resting place (presumably after build) will be a separate example repository. In that case there will be no submodule at all.

There are legal considerations in redistributing 3rd party code like this. I have updated my previous comment to include some of them.

File: orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
+import onnx
```
Comment (KeDengMS) [inline orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py:1]: should we add copyright headers?

File: orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        return args.learning_rate
+
+    schedule_fct = SCHEDULES[schedule]
+    return args.learning_rate * schedule_fct(training_steps / args.max_steps, args.warmup_proportion)
+
+def run_ort_training_step(args, global_step, training_steps, model, batch):
+    input_ids, segment_ids, input_mask, masked_lm_labels, next_sentence_labels = batch
+
+    if args.fp16:
+        loss_scaler = LossScaler(model.loss_scale_input_name, True, up_scale_window=2000)
```
Comment (Tixxx) [inline orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py:110]: There's another issue here. This means every step we will create a new lossScalar with a large magnitude, which means the gradients will keep overflowing and weights will never be updated. I will suggest moving this to create_ort_trainer and add it as an attr to args, i.e.:
```
if args.fp16:
    setattr(args, 'ort_loss_scale', LossScaler(model.loss_scale_input_name, True, up_scale_window=2000))
return model
```
Then in run_ort_training_step you can just do
```
loss_scale = torch.tensor([args.ort_loss_scale.loss_scale_])
```
and
```
if args.fp16:
    args.ort_loss_scale.update_loss_scale(all_finite.item())
```

**Comment by suffiank** [comment]
Will create a new PR using branch https://github.com/microsoft/onnxruntime-training-examples/tree/sukha/nvbert against examples repository.

**Comment by suffiank** [comment]
OK! Moved to https://github.com/microsoft/onnxruntime-training-examples/pull/2

File: orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py (hunk 10 above)
Comment (suffiank) [reply orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py:110]: I made these fixes in PR at https://github.com/microsoft/onnxruntime-training-examples/pull/2

File: orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py (hunk 9 above)
Comment (suffiank) [reply orttraining/pytorch_frontend_examples/nvidia-bert/ort_supplement/ort_supplement.py:1]: Added copyrights to new PR https://github.com/microsoft/onnxruntime-training-examples/pull/2

**Comment by thiagocrepaldi** [comment]
Are we closing this PR?

**Comment by thiagocrepaldi** [comment]
Closed as it was moved to https://github.com/microsoft/onnxruntime-training-examples/pull/2
